### ISSUE TITLE
Update to JTS 1.16.0

### ISSUE THREAD

**Issue by airbreather** [NetTopologySuite/NetTopologySuite#265]
[JTS 1.16.0](https://github.com/locationtech/jts/releases/tag/jts-1.16.0)

There's at least one intentional breaking change in there (locationtech/jts@c709631 removed the `(int, int)` constructor that [we still have](https://github.com/NetTopologySuite/NetTopologySuite/blob/c94898f6a6282f6d8bb4fa65062ac59324108b85/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequenceFactory.cs#L52-L61)), so we would technically violate SemVer if we released it as NTS 1.16, but we have a history of way bigger SemVer violations, so it shouldn't be the end of the world.

Here's the list of commits to consider.

- [x] locationtech/jts@4d5b984 (JTS-specific, meta)
- [x] locationtech/jts@7da8065 (JTS-specific, meta)
- [x] locationtech/jts@2387a45 (entirely in GeoAPI)
- [x] locationtech/jts@d13f230 (f8f20d1)
- [x] locationtech/jts@c709631 (bba1334)
- [x] locationtech/jts@868c34b (00d97f4)
- [x] locationtech/jts@8c5db26 (ce13e8d1)
- [x] locationtech/jts@8ef27d4 (5161d65)
- [x] locationtech/jts@e302001 (18aad66)
- [x] locationtech/jts@06e0424 (ed5be57)
- [x] locationtech/jts@17e9b05 (d00fed1)
- [x] locationtech/jts@bd4ca6e (ca328e4)
- [x] locationtech/jts@412bf17 (JTS-specific, meta)
- [x] locationtech/jts@7f8cde4 (JTS-specific, meta)
- [x] locationtech/jts@b1422b4 (2878d59)
- [x] locationtech/jts@2477de9 (1c9cded)
- [x] locationtech/jts@0b75fe7 (d1a5ed1)
- [x] locationtech/jts@19e6f32 (82acbe3)
- [x] locationtech/jts@b464d54 (f4afb40)
- [x] locationtech/jts@58ba543 (52d1f7e)
- [x] locationtech/jts@09e3554 (9715e0d)
- [x] locationtech/jts@ef26db0 (35cf416)
- [x] locationtech/jts@88e967b (0d5386c)
- [x] locationtech/jts@a1009c1 (e219d3c)
- [x] locationtech/jts@3221a31 (nothing to do, we've always used the same GeoAPI enum)
- [x] locationtech/jts@8d433ca (JTS-specific, meta)
- [x] locationtech/jts@f72773a (655db33)
- [x] locationtech/jts@7d7d796 (602f10d)
- [x] locationtech/jts@83b684f (0331f76)
- [x] locationtech/jts@34cf94d (JTS-specific, meta)
- [x] locationtech/jts@d3d91ab (JTS-specific, meta)

**Comment by airbreather** [comment]
```
C:\SNIP\jts>git log --reverse --pretty=oneline bac855f..d3d91ab
4d5b9845c5492cddcfbfe4be9ade49d689346a0d [maven-release-plugin] prepare for next development iteration
7da8065738cac413b018dba2ddac4257c0ed63ca Fix references to 1.15-SNAPSHOT
2387a4563d6803dacdbe572409d264bd62f98cbf clean up java comparable and comparator warnings
d13f2309c3c28eb3aa0e28d19624fb4d48240914 Add CoordinateSequence.getNumberOfMeasures() default method
c709631e2de44c67306d790eb9ef9c8327cce3bf Coordiante accessor methods for getX, getY, getZ, getM
868c34bc3f899413f4b4ae7118d2354edc1b7145 Introduce CoordinateXY, CoordinateXYM, CoordinateXYZM
8c5db263b2316c363c7fe7108585d0621fe04a6e improved javadocs
8ef27d434edbdebf96d9b61be1810319a435ae7d test coordinate subclasses and packed coordinate sequence
e30200157710cd15c7e338f42a512322a371134c confirm using factory to copy coordiante sequence works as expected
06e04243d58423f0a71401f4f2eeb06e4f09cf9b Updated CoordinateArraySequence to support measure
17e9b05e88f82ba6de0d6448a6e8b5b4c9711d76 ExtendedCoordinateSequence correctly reports measure
bd4ca6ec1516d99aee727a0a16ef27230fe8760f revise based on feedback
412bf173c241ed08b6862eae393a79f56f15c8da address typos and header feedback
7f8cde4a5c0d7132e50dbed9ffad832683bb2e99 change to 1.16.0-SNAPSHOT
b1422b43ac05444373d7eea059b8c07290e3b222 allow subclasses to change dimension and measure on the fly
2477de9d31bf4ccf8955d069a8a1c3c534079e19 Fix use of Coordinate copy and setCoordinate methods
0b75fe7471dabccb32557a175b8f31d3af1d4a57 Introduce enforceArrayConsistency check for CoordinateArraySequence
19e6f3234a318fd9806dbf5e96fe6699262f0bcf Restore PackedCoordinateSequence create( array, dimension ) methods
b464d54960cf0c99195a0f93a26ccfeab55c7f55 Changed WKTReader/-Writer to work on CoordinateSequences Changes to WKTReader * added ALLOW_OLD_JTS_COORDINATE_SYNTAX configuration, including   member variable that can be overridden * added member variable to ALLOW_OLD_JTS_MULTIPOINT_SYNTAX * added `measureToZ` flag value, that indicates for 'GEOMETRY M'   entities the measure value is stored as z-ordinate value. * cleaned up class
58ba543048998f4914bb8461dadcaa943b0c13d1 Improvements to WKTReader/-Writer * WKTReader   * moved ordinate flags from `WKTReader` to `CoordinateSequence` * WKTWriter   * now controlling `WKTWriter` output with ordinate flags   * added OutputOrdinates property getter and setter   * added unit test for `WKTWriter` properties
09e35542f91fc9d557ceced800cd57bd3d94ba16 * Fix mergeSequences for z- and m-ordinates * Renamed old WKTReaderTest -> WKTReadWriteTest * Added WKTReaderTest that compares the coordinate sequences of the   geometries for equality * Added lots of utility methods to GeometryTestCase, made it abstract
ef26db0131326c5a6375d89f69f9e2dbb5da6791 Fix WKTReadWriteTest by setting output ordinates to XY
88e967b19c824bf146bf95caa008c0831dd0d8be Fixing GeometryFactory createMultiPoint(CoordinateSequence) to honor possible measure values
a1009c1c9ca612447c97a747ca246ddbd0dfbf80 Adjusting PR to honer latest changes to CoordinateSequence in respect to XYZM support. - removed ordinate flag definitions from CoordinateSequence. - added jts.IO.WKTOrdinates enum class - shifted logic from ordinate flags to WKTOrdinates enum. - fixed tests
3221a317dda5035b1312ed9dbfac040a092cc01d Refactored WKTOrdinates to Ordinates
8d433caaf37d0dd965af95608c23acbb9dbdb566 (HEAD -> master, origin/master, origin/HEAD) update version history
f72773a8e18ee90451e21e9373ed5cecb456789b Revert use of string.format (#309)
7d7d796cc3f1c594cacf3d8129dbc0ec3ba4601a Minor doc fix
83b684f0a93ce1034bd4f40b6bba3c73745357bc Test case confirming factory dimension clipping behaviour (#316)
34cf94dc316094cc1ef515169f94f0299d0c70d1 Adding some details about how to publish to Maven Central. (#299)
d3d91aba0c81d38fd80e172524aa9e2385db87f9 (tag: jts-1.16.0) [maven-release-plugin] prepare release jts-1.16.0
```

**Comment by airbreather** [comment]
Ugh, PackedCoordinateSequence is a mess.  NTS / GeoAPI coordinate sequences are different enough from JTS's that it might not be worth trying to preserve strict parity between the different lines, and so maybe it's worth rewriting even more idiomatic to .NET.

For example, NTS lets you store and retrieve any random assortment of **well-defined** ordinate values... or at least it lets you try to (it's not actually working properly :cry:):
- NTS (if it were implemented correctly) could let you store anything... sure, any of XY, XYZ, XYM, XYZM, but also XYZ plus Ordinate23, or XY plus Ordinate5.
- with JTS, it just has two distinctions: "how many values are present per coordinate" and "how many of that number of values are 'measures' (i.e., not spatial dimensions)"
- so on the JTS side, looking at just spatial dimensions (i.e., all dimensions that aren't "measures"), it looks like you could only have XY, XYZ, XYZD, XYZDD, XYZDDD, where each extra D is some other spatial dimension... in other words, going the JTS way, each additional spatial dimension can only be present if all previous spatial dimensions are present as well.  this makes logical sense, I suppose, but hold on...
- still on the JTS side, turning now to just "measures", it looks like you can only have
    - no M
    - M<sub>1</sub>
    - M<sub>1</sub> and M<sub>2</sub>
    - M<sub>1</sub>, M<sub>2</sub>, and M<sub>3</sub>
- in other words, even though JTS gives you the facilities to store as many "measures" as you can define, if you're actually going to **use** more than one, those same facilities don't let you say "this sequence's single measure is, say, 'distance in meters', whereas this other sequence's single measure is, say, 'freeway exit numbers'".  the only way to unambiguously define that is by carrying extra data in "lower" measures such that the first measure is present, but populated with NaN, so that you can have the second measure.

Example of what I think we should be able to do with NTS / GeoAPI, given the way that GeoAPI's interfaces use enums (sometimes with flags) to describe the data:
```csharp
var pdcs = BuildMyPackedDoubleCoordinateSequenceWithFreewayExitNumbers();
Assert.IsTrue(pdcs.Ordinates == Ordinates.XYZ | Ordinates.Ordinate10);
var firstX = pdcs.GetOrdinate(0, Ordinate.X);
var firstY = pdcs.GetOrdinate(0, Ordinate.Y);
var firstZ = pdcs.GetOrdinate(0, Ordinate.Z);
var firstFreewayExitNumber = pdcs.GetOrdinate(0, Ordinate.Ordinate10);

var pdcs2 = BuildMyPackedDoubleCoordinateSequenceWithLinearDistances();
Assert.IsTrue(pdcs2.Ordinates == Ordinates.XYZ | Ordinates.Ordinate11);
var firstX = pdcs2.GetOrdinate(0, Ordinate.X);
var firstY = pdcs2.GetOrdinate(0, Ordinate.Y);
var firstZ = pdcs2.GetOrdinate(0, Ordinate.Z);
var firstLinearDistance = pdcs2.GetOrdinate(0, Ordinate.Ordinate11);
```

... ... ... either that, or we shouldn't bother with all this nonsense, and just stick with XY / XYZ / XYM / XYZM since that's all that can reasonably possibly work with the rest of the system.

**Comment by airbreather** [comment]
> ... ... ... either that, or we shouldn't bother with all this nonsense, and just stick with XY / XYZ / XYM / XYZM since that's all that can reasonably possibly work with the rest of the system.

Side note, even the existing NTS PackedCoordinateSequence stuff seems like it would fail with XYM, since we'll try to go to the same spot that we would with XYZM (and there's a [comment](https://github.com/NetTopologySuite/GeoAPI/blob/5c1628cbb769aa3604ef3577eac6109dccb9f743/src/GeoAPI/Geometries/Ordinate.cs#L32-L35) in there to that effect).

If we want to be able to nicely store both XYM and XYZM, then the same tricks to enable that would also enable the `Ordinates.Ordinate10` / `Ordinates.Ordinate11` distinction, which is why I point it out.

**Comment by airbreather** [comment]
EHH.  No matter which way we go with the current roadblock, updating to JTS 1.16.0 and higher will require breaking changes, if for no other reason than because GeoAPI is also taking some big breaking changes.

I have a clever replacement for Packed*CoordinateSequence that ignores its JTS heritage and focuses on what's efficient on .NET, but it depends on stuff that's only present in .NET Standard 1.1 / .NET Framework 4.5 and above, so we'd have to leave even more target frameworks behind to get it in.

Because of the breaking changes. should we take this opportunity to make the next version of NTS (and perhaps GeoAPI as well?) **only** target .NET Standard 2.0 (maybe also .NET 4.6.1 [per Microsoft's guidance](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting#multi-targeting))?
- .NET Standard targets below 2.0 are [discouraged](https://docs.microsoft.com/en-us/dotnet/standard/library-guidance/cross-platform-targeting#net-standard).
- Compared to .NET Standard 2.0, continuing to support .NET Framework 4.5 would continue to limit our choices for the foreseeable future until we take yet another breaking change.
- If we don't have to build separate .NET Framework targets, then Travis-CI builds can stop fetching Mono every time (slow!), and everything just generally gets a little bit easier to build on Linux.

I wish NuGet.org gave us detailed stats so we could know how much this is used on .NET Framework <4.6.1, since those are the only people who would be affected...

**Comment by airbreather** [comment]
I suppose it therefore follows that the members exposed on GeoAPI's `ICoordinateSequence` / `ICoordinateSequenceFactory` should be considered potentially non-final as well.

At a quick glance, several `Create` overloads on the factory seem redundant with the one that accepts a length and `Ordinates` flags... I may be wrong, though.

**Comment by DGuidi** [comment]
> I have a clever replacement for Packed*CoordinateSequence that ignores its JTS heritage and focuses on what's efficient on .NET, but it depends on stuff that's only present in .NET Standard 1.1 / .NET Framework 4.5 and above, so we'd have to leave even more target frameworks behind to get it in.

The biggest concern to me is to take a road of separation from JTS code, a road that can be difficult to maintain in the future when JTS code changes, but I trust in your confidence with JTS and NTS code.
+1 for breaking changes, if code and performances are better: for semver, I'd like to match JTS version anyway.

> Because of the breaking changes. should we take this opportunity to make the next version of NTS (and perhaps GeoAPI as well?) **only** target .NET Standard 2.0 

+1 for this. I see that **a lot** of libraries are moved to net standard and cleaned their API as well.

**Comment by airbreather** [comment]
> The biggest concern to me is to take a road of separation from JTS code, a road that can be difficult to maintain in the future when JTS code changes, but I trust in your confidence with JTS and NTS code.

If it were just "do we start diverging or not?", then I wouldn't even bring it up.  **NTS has already diverged from JTS**: `ICoordinateSequence` lets you ask for values at well-defined ordinates (e.g., "the Z-ordinate", "the M-ordinate", "the 23rd ordinate"), whereas JTS (pre-1.16) only let you ask for either:
- X / Y (via `getX()` / `getY()`), or
- the value of the "`n`th" ordinate

NTS's packed sequence has always (since the [very first version](https://github.com/NetTopologySuite/NetTopologySuite/blob/df7c4296c236c888910ce49401b73c6955fc90ba/NetTopologySuite/Geometries/PackedCoordinateSequence.cs#L352)) actually *worked* the same way, but the public API has made it *appear* like you could do things that you couldn't do in JTS, but which would actually never work (i.e., distinguish between XYM / XYZ sequences without external information).

Starting with 1.16.0, JTS's sequence still works mostly the same way it always has, *however*, it expanded that first bullet point to include Z (via `getZ()`) and M (via `getM()`).  The interesting part is that JTS's `getZ()` on an XYZ sequence and `getM()` on an XYM sequence will both get the ordinate at index 2.
- In the JTS world, that makes perfect sense: it uses an integer for the ordinate index.
- In the NTS world, it's crazy, because "index 2" is not just `Ordinate.Ordinate2`, but also `Ordinate.Z`.

Another way to put it is that if we don't do something radical, `seq.GetM(4)` would sometimes give you something different than `seq.GetOrdinate(4, Ordinate.M)`, which I think is unacceptable in the version where we start saying that we more completely support Z and M.

So the decision point isn't quite "diverge or don't diverge?", but rather one of:
1. keep this misleading API around, with all the same warts that it's always had, just adding the direct counterparts to what JTS added for 1.16.0,
1. make NTS look **more** like JTS 1.16.0: everything flows from `Dimension` and `Measures`, and coordinate sequences no longer use `Ordinate` / `Ordinates` except where JTS does, or
1. do a clean separation from JTS at this layer, and have everything flow from `Ordinate` / `Ordinates` (probably remove `Dimension` and `Measures` too: `Ordinates` tells the same story more completely).

I lean strongly towards that last bullet point there.  I don't really think the JTS `ordinateIindex` is actually an important concept to the users apart from performance... and for performance, folks could still use `GetRawCoordinates()` to get mostly the same thing without us tarnishing the `ICoordinateSequence` API.

> if code and performances are better

[This](https://github.com/NetTopologySuite/NetTopologySuite/blob/b8427723a63f2dc03f7dd8ee210d61a249d673f3/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequence.cs) is likely to be a challenge to maintain, though I think it's likely to be the fastest way to do this.  Note that all those type checks happen **at JIT time**, so the machine code should be the best we can hope for.

A "safer" alternative I've been considering was to read the `Ordinates` flags and then build a mapping of ordinate to packed index, but that looked just about as obnoxious, and I **suspect** that performance would be somewhat worse when reading Z and M because the JIT doesn't know ahead of time where to look for them, or if they'll even be there at all.

**Comment by FObermaier** [comment]
The `Ordiante` (?) and `Ordinates` enum have been developed in attempt to make NTS more .NET like. The attempt failed, but I thought not all of it is bad and brought those enums in the main branch.
I tried to persuade JTS' folks to follow that lead, but they chose not to. That's why we are at this point now.

So, I have no problems with ditching `Ordinates`. As for `Ordinate`: JTS does have a [definition of standard ordinate indices](https://github.com/locationtech/jts/blob/62dbb71a471fadea6ebda3b42f948d821a12f293/modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java#L48-L69). Nothing else is `Ordinate` enum.

I strongly support having additional/better `ICoordinateSequence` implementations, there is absolutely nothing standing in the way. Additionally I think that it is feasible to use value coordinates internally (structs) for computation if that can improve the overall performance. Memory<T>/Span<T>

**Comment by airbreather** [comment]
> So, I have no problems with ditching `Ordinates`. As for `Ordinate`: JTS does have a [definition of standard ordinate indices](https://github.com/locationtech/jts/blob/62dbb71a471fadea6ebda3b42f948d821a12f293/modules/core/src/main/java/org/locationtech/jts/geom/CoordinateSequence.java#L48-L69). Nothing else is `Ordinate` enum.
> 
> I strongly support having additional/better `ICoordinateSequence` implementations, there is absolutely nothing standing in the way. Additionally I think that it is feasible to use value coordinates internally (structs) for computation if that can improve the overall performance. Memory/Span

OK, I'll work on making GeoAPI's `ICoordinateSequence` and its implementations in NTS match JTS as closely as possible.  I wouldn't mind putting some other faster implementations in there, later.

For `GetOrdinate` / `SetOrdinate`, I propose that we should either:
- match JTS by taking in `int ordinateIndex`, or
- at least remove `Ordinate.Z` and `Ordinate.M` from the enum, making it { X, Y, Ordinate2, Ordinate3, ... }

AFAIK, the main reason to use `enum` instead of `int`, in .NET anyway, is to remove the need to consider the actual integer value of whatever you're looking at.  With `GetOrdinate(3, Ordinate.M)` sometimes returning a different value than `GetM(3)`, I feel like we're failing at that, even if this behavior is documented on the enum value itself.

In other words, it should be the method's responsibility to know what the user is trying to get / set when they pass in `Ordinate.M`, not the user's responsibility to know the circumstances where `Ordinate.M` is or is not the way to get / set an M value.

I also kinda disagree with JTS's decision to even have `CoordinateSequence.Z` / `CoordinateSequence.M` values for basically the same reasons, but not quite as strongly.

**Comment by airbreather** [comment]
What's the story behind all the discrepancies between our WKBReader and JTS's?

Any objections to bringing that closer to JTS as well?

**Comment by FObermaier** [comment]
AFAIK it has always been different than JTS'. If I'm not mistaken @DGuidi has included it from a different source?

**Comment by DGuidi** [comment]
As far as I can remember, IO stuff comes from [GeoTools.NET](http://geotoolsnet.sourceforge.net/Index.html) project, but I'm pretty sure that WKB/WKT stuff was NOT copied/adapted from this library, but was instead _inspired_ to JTS implementation...

Anyway, no objection to make this code closer to JTS

**Comment by airbreather** [comment]
I think I'm still getting too carried away with this, either way.  I keep trying to make the API match JTS (justifying it by saying "well that's the main feature of JTS 1.16 so it counts"), and it keeps leading me into places that I probably shouldn't be right now.

I'm going to switch gears a bit and just try to make sure that all the JTS 1.16 stuff has an equivalent in NTS (even if NTS already had something for that which looked different), and then make sure to open some new issues with next-steps to consolidate / discuss / improve issues I find.

**Comment by DGuidi** [comment]
Let me clarify a bit what I meant when I talked about "make NTS as close as possible to JTS". To me, when in the past years I mantained this lib almost alone, in my free time and without any background in _topology stuff_, the biggest values are:

1. **speed** adding new features: almost copy code, fix compilation issues (and avoid replace `Equals `calls with `==` to save days of debugging headataches)
1. **debugging**: if someone posted a bug of a strange behavior, the only thing I do was to write the same test code in java and see what happens; most of the time, NTS behavior was exact the same as JTS, so my response was pretty straightforward "post this issue to JTS mailing list using this java code as example"

So basically my suggests come from that experience.

Now that NTS code is maintained mainly by you and felix (with help from other people too), experienced programmers with a strong GIS background, feel free to make the changes you want :)

**Comment by airbreather** [comment]
> Now that NTS code is maintained mainly by you and felix (with help from other people too), experienced programmers with a strong GIS background, feel free to make the changes you want :)

Of course, but there's still a lot of value in keeping things very close to JTS, at least when it comes to the API.

It's not just JTS / NTS, but also JSTS, GEOS, and miscellaneous other language bindings using GEOS as the engine.  For the most part, I think all these libraries have the same basic API as JTS, just in their own flavor.

So for example, if someone works on, say, a rich ASP.NET Core application that uses NTS, they can choose to push the WKB down and manipulate it client-side using JSTS, with essentially the same tools.  I feel that this familiarity counts for a lot.

Another benefit: if we have the exact same implementations (within reason), then bugs found by our own community will often apply to everyone else, and so we can report them upstream and have all these other communities benefit from the fix.

There are drawbacks, of course:
1. Speed / memory usage.  The .NET and Java runtimes are different in ways that likely matter.
1. It's risky to try to fix JTS problems in NTS, because then if JTS fixes those same problems in a different way, we have to figure out which version(s) of the fix to continue supporting, and how.
1. Java and .NET don't support exactly the same sets of language features.

At least those last two have hit as I've been working on this, and especially the second one is why I've been coming up with some unusual ideas on this thread that, in the ordinary course of things, would be insane to even bring up, let alone seriously consider.

Despite those drawbacks, I completely agree that that NTS should match JTS as closely as reasonably possible, **at least** in the public API.

**Comment by airbreather** [comment]
Lots of new test failures (about 100) when updating WKTReader for locationtec

I picked one at random... it brought me to the NTS equivalent of [this](https://github.com/locationtech/jts/blob/b464d54960cf0c99195a0f93a26ccfeab55c7f55/modules/core/src/main/java/org/locationtech/jts/algorithm/Length.java#L40), which is technically a bug on the JTS side of things (if `pts` is an XYZM sequence, then this would throw the same way that XYZ sequences throw in NTS).

I'm going to press on ignoring these additional test failures (and related ones that came before them for apparently related reasons) and then fix them up after this series.

**Comment by FObermaier** [comment]
@airbreather , [locationtech/jts@b464d54](https://github.com/locationtech/jts/commit/b464d54) belongs to a series of commits. If the problems still exist with [locationtech/jts@88e967b](locationtech/jts@88e967b) we should care.

**Comment by airbreather** [comment]
> If the problems still exist with locationtech/jts@88e967b we should care.

Going through more of them, the problems are overwhelmingly the same: `System.InvalidOperationException : Coordinate does not support setting Z-ordinate`

Each place I dig into, there's either a test that's expecting `Coordinate.Z` to work despite that test creating an XY geometry (so a bad test), or there's something in JTS that's doing `new Coordinate()` instead of `someCoordinateSequence.CreateCoordinate()` (so bad JTS code).

Both stem from the decision we made around the time of NetTopologySuite/GeoAPI#55 to have the base `Coordinate` class no longer support Z (despite JTS's `Coordinate` continuing to work with Z), not anything to do with the JTS port.  I've fixed *just* enough of these discrepancies to make the solution build (e.g., `Coordinate.EqualInZ` became our `CoordinateZ.EqualInZ`), but there are certainly more of these that the JTS port alone won't fix.

I expect that there might be a few more that crop up in the remaining commits, so at the end of the 1.16.0 series, my plan is to categorize the test failures and fix them accordingly all at once:

|Failure Reason|Fix|
|-|-|
|I made a mistake in porting some commit|Rewrite the commit to fix the issue|
|Test expects Z to work for an XY geometry|Change the test to make an XYZ geometry, or to stop trying to put Z where it doesn't belong|
|Code in NTS proper is trying to do something with `Coordinate` that's actually correct in JTS|Change it to use `CoordinateZ`|
|Code in NTS proper is trying to do something with `Coordinate` that's also broken in JTS|Correct it in NTS (if possible), and add it to an upstream bug report that logs all these at once|
|Something else|Case-by-case|

**Comment by FObermaier** [comment]
@airbreather, if there is anything you want me to have a look at, let me know.
Your planned failue case approach looks reasonable to me.

**Comment by DGuidi** [comment]
NOTE: **TOTALLY** unrelated to this work, maybe

I'm not sure if this can be helpful, but reading [Is C# a low-level language?](https://mattwarren.org/2019/03/01/Is-CSharp-a-low-level-language/) I readed something maybe interesting:

> As of ‘.NET Standard 2.1’ there are now specific float implementations of the common maths functions, located in the System.MathF class.

Performance improvements are promising, not sure if single precision is enough for NTS algorithms: maybe can be used based on geometry factory configuration?

**Comment by airbreather** [comment]
> > As of ‘.NET Standard 2.1’ there are now specific float implementations of the common maths functions, located in the System.MathF class.
> 
> Performance improvements are promising, not sure if single precision is enough for NTS algorithms: maybe can be used based on geometry factory configuration?

The immediate problem that I have with anything that targets .NET Standard 2.1 is that the crusty ol' .NET Framework has no plans to implement it in the foreseeable future.  The application I maintain at work uses a **ton** of stuff that's incompatible with .NET Core (even considering what's coming in .NET Core 3.0), and I suspect that this is far from the only dependent that's stuck on .NET Framework for the time being.

Specifically with NTS and the idea of using `System.MathF`, I doubt we could see enough improvement to justify maintaining a separate stack for single-precision operations.

- I haven't profiled in a while, but IIRC, the limiting part of overlay (the slowest operation that I know of in NTS) is sorting the list of sweep-line events, which is independent of single- vs. double-precision.
- There's more potential with point-in-polygon, but after using an R-tree to index segments by Y-value, I've never seen it slow enough to justify significant improvements.

**Comment by airbreather** [comment]
All that said, this is a great article that touches on a lot of features that we can look into using to improve our performance.  [The crazy `ICoordinateSequence` implementation](https://github.com/NetTopologySuite/NetTopologySuite/blob/b8427723a63f2dc03f7dd8ee210d61a249d673f3/NetTopologySuite/Geometries/Implementation/PackedCoordinateSequence.cs) I linked earlier in this thread was an example of what some of it might look like.  It's like looking at a completely different language in some places.

**Comment by airbreather** [comment]
OK, all the individual commits have made it in.  Now to fix the test failures...

**Comment by airbreather** [comment]
JTS bug I mentioned earlier was actually partially opened already... I've just added a comment here: https://github.com/locationtech/jts/issues/370#issuecomment-468926825
